pkg/controller/expand_and_recover_test.go

```diff
@@ -19,76 +19,86 @@ import (
 	fsVolumeMode := v1.PersistentVolumeFilesystem
 	var tests = []struct {
-		name                 string
-		pvc                  *v1.PersistentVolumeClaim
-		pv                   *v1.PersistentVolume
-		recoverFeatureGate   bool
-		disableNodeExpansion bool
+		name                       string
+		pvc                        *v1.PersistentVolumeClaim
+		pv                         *v1.PersistentVolume
+		disableNodeExpansion       bool
+		disableControllerExpansion bool
 		// expectations of test
 		expectedResizeStatus  v1.PersistentVolumeClaimResizeStatus
 		expectedAllocatedSize resource.Quantity
 		expectResizeCall      bool
 	}{
 		{
-			name:                  "pvc.spec.size > pv.spec.size, recover_expansion=on",
+			name:                  "pvc.spec.size > pv.spec.size, resize_status=node_expansion_inprogress",
 			pvc:                   getTestPVC("test-vol0", "2G", "1G", "", v1.PersistentVolumeClaimNoExpansionInProgress),
 			pv:                    createPV(1, "claim01", defaultNS, "test-uid", &fsVolumeMode),
-			recoverFeatureGate:    true,
 			expectedResizeStatus:  v1.PersistentVolumeClaimNodeExpansionPending,
 			expectedAllocatedSize: resource.MustParse("2G"),
 			expectResizeCall:      true,
 		},
 		{
-			name:                  "pvc.spec.size = pv.spec.size, recover_expansion=on",
+			name:                  "pvc.spec.size = pv.spec.size, resize_status=no_expansion_inprogress",
 			pvc:                   getTestPVC("test-vol0", "1G", "1G", "", v1.PersistentVolumeClaimNoExpansionInProgress),
 			pv:                    createPV(1, "claim01", defaultNS, "test-uid", &fsVolumeMode),
-			recoverFeatureGate:    true,
 			expectedResizeStatus:  v1.PersistentVolumeClaimNodeExpansionPending,
 			expectedAllocatedSize: resource.MustParse("1G"),
 			expectResizeCall:      true,
 		},
 		{
-			name:                  "pvc.spec.size = pv.spec.size, recover_expansion=on",
+			name:                  "pvc.spec.size > pv.spec.size, resize_status=controller_expansion_failed",
+			pvc:                   getTestPVC("test-vol0", "5G" /*specSize*/, "3G" /*statusSize*/, "10G" /*allocatedSize*/, v1.PersistentVolumeClaimControllerExpansionFailed),
+			pv:                    createPV(3, "claim01", defaultNS, "test-uid", &fsVolumeMode),
+			expectedResizeStatus:  v1.PersistentVolumeClaimNodeExpansionPending,
+			expectedAllocatedSize: resource.MustParse("5G"),
+			expectResizeCall:      true,
+		},
+		{
+			name:                       "pvc.spec.size = pv.spec.size, resize_status=node_expansion_failed, disable_controller_expansion=true",
+			pvc:                        getTestPVC("test-vol0", "5G" /*specSize*/, "3G" /*statusSize*/, "10G" /*allocatedSize*/, v1.PersistentVolumeClaimNodeExpansionFailed),
+			pv:                         createPV(10, "claim01", defaultNS, "test-uid", &fsVolumeMode),
+			disableControllerExpansion: true,
+			expectedResizeStatus:       v1.PersistentVolumeClaimNodeExpansionPending,
+			expectedAllocatedSize:      resource.MustParse("5G"),
```

Comment:
This is odd... how can be AllocatedSize shrunk after failed node expansion? PV size is 10Gi already.

Reply:
Never mind, I see disableControllerExpansion. The test is correct.